In [1014]:
import pandas as pd
import numpy as np
import sidetable
import sys
import re
sys.path.append('../')

import seaborn as sns
import matplotlib.pyplot as plt

In [1015]:
import datetime

In [1016]:
pd.options.display.max_columns = None

In [1017]:
df_guiris = pd.read_csv(r"C:\Users\mituc\Ironhack\Curso\IronLabs\Proyecto-1\data\2078.csv", sep=';', encoding='latin-1', index_col = 0)

In [ ]:
df_guiris.shape

In [1018]:
df_guiris.head(5)

Viajeros y pernoctaciones             Residencia  \
Puntos turísticos                                                    
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en España   

                       Periodo   Total  
Puntos turísticos                   
01059 Vitoria-Gasteiz  2022M01  11.759  
01059 Vitoria-Gasteiz  2021M12  16.496  
01059 Vitoria-Gasteiz  2021M11  16.667  
01059 Vitoria-Gasteiz  2021M10  21.467  
01059 Vitoria-Gasteiz  2021M09  18.647

In [1019]:
df_guiris.duplicated().sum()

12322

In [1020]:
df_guiris[df_guiris.duplicated() == True]

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero        Residentes en España   
02003 Albacete                          Viajero  Residentes en el Extranjero   
02003 Albacete                          Viajero  Residentes en el Extranjero   
...                                         ...                          ...   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
02003 Albacete         2020M04     0  
02003 Albacete         2020M06     .  
02003 Albacete         2020M05     .  
...                        ...   ...  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M06     .  
50297 Zaragoza         2020M05     .  
50297 Zaragoza         2020M04     0  

[12322 rows x 4 columns]

In [1021]:
df_guiris[df_guiris['Total'] == '0']

Viajeros y pernoctaciones                   Residencia  \
Puntos turísticos                                                          
01059 Vitoria-Gasteiz                   Viajero        Residentes en España   
01059 Vitoria-Gasteiz                   Viajero  Residentes en el Extranjero   
01059 Vitoria-Gasteiz            Pernoctaciones        Residentes en España   
01059 Vitoria-Gasteiz            Pernoctaciones  Residentes en el Extranjero   
02003 Albacete                          Viajero        Residentes en España   
...                                         ...                          ...   
49275 Zamora                     Pernoctaciones  Residentes en el Extranjero   
50297 Zaragoza                          Viajero        Residentes en España   
50297 Zaragoza                          Viajero  Residentes en el Extranjero   
50297 Zaragoza                   Pernoctaciones        Residentes en España   
50297 Zaragoza                   Pernoctaciones  Residentes en el Extranjero   

                       Periodo Total  
Puntos turísticos                 
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
01059 Vitoria-Gasteiz  2020M04     0  
02003 Albacete         2020M04     0  
...                        ...   ...  
49275 Zamora           2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  
50297 Zaragoza         2020M04     0  

[425 rows x 4 columns]

Aunque parecía que había duplicados, en la tabla anterior podemos ver que no es así y si resetamos el index desaparecen

In [1022]:
df_guiris.reset_index(inplace=True)

In [1023]:
df_guiris.duplicated().sum()

0

In [1024]:
df_guiris.dtypes

Puntos turísticos        object
Viajeros y pernoctaciones    object
Residencia                   object
Periodo                      object
Total                        object
dtype: object

Tendremos que limpiar algunos nombres

In [1025]:
df_guiris.columns

Index(['Puntos turísticos', 'Viajeros y pernoctaciones', 'Residencia',
       'Periodo', 'Total'],
      dtype='object')

In [1026]:
print(df_guiris['Viajeros y pernoctaciones'].unique())
print(df_guiris['Residencia'].unique())

['Viajero' 'Pernoctaciones']
['Residentes en España' 'Residentes en el Extranjero']


In [1027]:
df_guiris.isnull().sum()

Puntos turísticos           0
Viajeros y pernoctaciones       0
Residencia                      0
Periodo                         0
Total                        5358
dtype: int64

In [1028]:
df_guiris.shape

(86920, 5)

Renombramos las columnas

In [1030]:
df_guiris = df_guiris.rename(columns={'Puntos turísticos': 'Puntos_turisticos'})
columnas_limpias = {col : col.lower().replace(" ", "_") for col in df_guiris.columns}
df_guiris.rename(columns = columnas_limpias, inplace = True)

In [1031]:
df_guiris.columns

Index(['puntos_turisticos', 'viajeros_y_pernoctaciones', 'residencia',
       'periodo', 'total'],
      dtype='object')

Ahora vamos a dejar limpita la columna 'total'. Primero retiraremos los puntos y luego convertiremos los valores en enteros.

Es importante tener en cuenta que en este csv si aparece un punto '.' a secas, significa que el dato está protegido por la ley de secreto estadístico, (Artículo 12 de la Ley 7/2006) y para nosotros es como si fuera un nulo.

Esta operación de limpieza tiene por tanto una doble función:
- poder convertir los string en int
- convertir los '.' en NaN para manejarlos después

In [1032]:
df_guiris['total'] = df_guiris['total'].str.replace('.', '')

C:\Users\mituc\AppData\Local\Temp/ipykernel_4944/2279084469.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df_guiris['total'] = df_guiris['total'].str.replace('.', '')


In [1033]:
df_guiris['total'] = pd.to_numeric(df_guiris['total'], errors='coerce').astype('Int64')

Podemos comprobar que los nulos han aumentado de 5358 a 13056

In [1034]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        13056
dtype: int64

Sabemos por la documentación del csv que entre abril y junio del 2020 no hubo recogida de datos a causa de la crisis del covid-19. Veamos cómo están esos datos.

In [1035]:
df_guiris[(df_guiris['periodo'] == '2020M04') | (df_guiris['periodo'] == '2020M05') | (df_guiris['periodo'] == '2020M06')]

puntos_turisticos viajeros_y_pernoctaciones  \
19     01059 Vitoria-Gasteiz                   Viajero   
20     01059 Vitoria-Gasteiz                   Viajero   
21     01059 Vitoria-Gasteiz                   Viajero   
224    01059 Vitoria-Gasteiz                   Viajero   
225    01059 Vitoria-Gasteiz                   Viajero   
...                      ...                       ...   
86530         50297 Zaragoza            Pernoctaciones   
86531         50297 Zaragoza            Pernoctaciones   
86734         50297 Zaragoza            Pernoctaciones   
86735         50297 Zaragoza            Pernoctaciones   
86736         50297 Zaragoza            Pernoctaciones   

                        residencia  periodo  total  
19           Residentes en España  2020M06   <NA>  
20           Residentes en España  2020M05   <NA>  
21           Residentes en España  2020M04      0  
224    Residentes en el Extranjero  2020M06   <NA>  
225    Residentes en el Extranjero  2020M05   <NA>  
...                            ...      ...    ...  
86530        Residentes en España  2020M05   <NA>  
86531        Residentes en España  2020M04      0  
86734  Residentes en el Extranjero  2020M06   <NA>  
86735  Residentes en el Extranjero  2020M05   <NA>  
86736  Residentes en el Extranjero  2020M04      0  

[1272 rows x 5 columns]

Dado que en los meses del confinamiento toda la actividad turística estuvo suspendida, vamos a sustituir los nulos por 0, pues no es tanto que no hubiese recogida de datos como que no había turistas.

In [1036]:
df_guiris.loc[(df_guiris['periodo'] == '2020M04') | 
              (df_guiris['periodo'] == '2020M05') | 
              (df_guiris['periodo'] == '2020M06'), 'total'] = 0

Y comprobamos que los nulos han descendido de 13056 a 12208

In [1037]:
df_guiris.isnull().sum()

puntos_turisticos                0
viajeros_y_pernoctaciones        0
residencia                       0
periodo                          0
total                        12208
dtype: int64

Los nulos que quedan se deben a que o bien no había recogida de datos en esa fecha para esa zona turística, o bien el dato no está publicado porque está protegido por secreto estadístico.
En todo caso no se pueden eliminar alegremente. En el caso de que falten en meses puntuales pero dispongamos de datos para el resto del año, quiáz convendría sustiruirlos.

In [1038]:
null_rows = pd.isnull(df_guiris['total'])

In [1039]:
df_guiris[null_rows].to_csv('nulls.csv')

In [1040]:
df_guiris['periodo'] = df_guiris['periodo'].str.replace('M', '-')

In [1041]:
df_guiris.to_pickle('guiris.pkl')
#df_guiris = pd.read_pickle('guiris.pkl')

In [1042]:
df_guiris.head(2)

puntos_turisticos viajeros_y_pernoctaciones             residencia  \
0  01059 Vitoria-Gasteiz                   Viajero  Residentes en España   
1  01059 Vitoria-Gasteiz                   Viajero  Residentes en España   

   periodo  total  
0  2022-01  11759  
1  2021-12  16496

In [1043]:
df_guiris['periodo'] = pd.to_datetime(df_guiris['periodo'], format='%Y%m%')
#df_guiris['periodo'] = df_guiris['periodo'].dt.date.apply(lambda x: x.strftime('%Y-%m'))

In [1044]:
df_guiris.dtypes

puntos_turisticos                    object
viajeros_y_pernoctaciones            object
residencia                           object
periodo                      datetime64[ns]
total                                 Int64
dtype: object

In [1045]:
df_guiris.set_index('periodo', inplace=True)

Renombramos los valores

In [1046]:
df_guiris['residencia'].replace({'Residentes en España': "residentes_espana", 'Residentes en el Extranjero': 'residentes_en_el_extranjero'}, inplace=True)
df_guiris['viajeros_y_pernoctaciones'] = df_guiris['viajeros_y_pernoctaciones'].str.lower()

In [1047]:
df_guiris.tail(5)

puntos_turisticos viajeros_y_pernoctaciones  \
periodo                                                  
2005-05-01    50297 Zaragoza            pernoctaciones   
2005-04-01    50297 Zaragoza            pernoctaciones   
2005-03-01    50297 Zaragoza            pernoctaciones   
2005-02-01    50297 Zaragoza            pernoctaciones   
2005-01-01    50297 Zaragoza            pernoctaciones   

                             residencia  total  
periodo                                         
2005-05-01  residentes_en_el_extranjero  20698  
2005-04-01  residentes_en_el_extranjero  20003  
2005-03-01  residentes_en_el_extranjero  21156  
2005-02-01  residentes_en_el_extranjero  13372  
2005-01-01  residentes_en_el_extranjero  10284

In [1091]:
def pastor_de_guiris(df, ciudades):
    for ciudad in ciudades:
        # Separamos la zona turistica
        df_guiris_zona = df[df['puntos_turisticos'].str.contains(ciudad) == True]
        # Aislamos los viajeros residentes en Espana
        df_numguiris = df_guiris_zona[df_guiris_zona['viajeros_y_pernoctaciones'].str.contains('viajero') == True]
        df_numguiris_esp = df_numguiris[df_numguiris['residencia'].str.contains('residentes_espana') == True]
        # Aislamos a los turistas residentes en el extranjero
        df_numguiris_ext = df_numguiris[df_numguiris['residencia'].str.contains('residentes_en_el_extranjero') == True]
        # Aislamos las pernocctaciones de los viajeros residentes en Espana
        df_pernocguiris = df_guiris_zona[df_guiris_zona['viajeros_y_pernoctaciones'].str.contains('pernoctaciones') == True]
        df_pernocguiris_esp =  df_pernocguiris[df_pernocguiris['residencia'].str.contains('residentes_espana') == True]
        # Aislamos las pernocctaciones de los viajeros residentes en el extranjero
        df_pernocguiris_ext =  df_pernocguiris[df_pernocguiris['residencia'].str.contains('residentes_en_el_extranjero') == True]

        guiris_esquilados = pd.concat([df_numguiris_esp['total'], 
                                    df_numguiris_ext['total'],
                                    df_pernocguiris_esp['total'],
                                    df_pernocguiris_ext['total']], axis=1)

        guiris_esquilados.columns = ['viajeros_espana', 
                                    'viajeros_extranjero',
                                    'pernoctaciones_espana',
                                    'pernoctaciones_extranjero']

        guiris_esquilados['media_pernoc_esp'] = df_pernocguiris_esp['total'] / df_numguiris_esp['total']
        guiris_esquilados['media_pernoc_ext'] = df_pernocguiris_ext['total'] / df_numguiris_ext['total']
        guiris_esquilados['total_viajeros'] = df_numguiris_esp['total'] + df_numguiris_ext['total']
        guiris_esquilados['total_pernoctaciones'] = df_pernocguiris_esp['total'] + df_pernocguiris_ext['total']
        guiris_esquilados['media_total_pernoc'] = guiris_esquilados['total_pernoctaciones'] / guiris_esquilados['total_viajeros']

        guiris_esquilados.to_pickle(f'{ciudad}_guiris_esquilados.pkl')
    
    return guiris_esquilados

In [1101]:
lista = ['Barcelona']

In [1102]:
pastor_de_guiris(df_guiris, lista)

viajeros_espana  viajeros_extranjero  pernoctaciones_espana  \
periodo                                                                   
2022-01-01           107162               194312                 203683   
2021-12-01           144299               250133                 282398   
2021-11-01           120748               314677                 228675   
2021-10-01           130404               364349                 270494   
2021-09-01           102053               309572                 207329   
...                     ...                  ...                    ...   
2005-05-01           118498               274735                 283763   
2005-04-01           118215               227402                 265758   
2005-03-01            98475               209654                 223927   
2005-02-01            91128               184467                 206601   
2005-01-01            85815               151746                 174598   

            pernoctaciones_extranjero  media_pernoc_esp  media_pernoc_ext  \
periodo                                                                     
2022-01-01                     541541          1.900702          2.786966   
2021-12-01                     674630          1.957034          2.697085   
2021-11-01                     871021           1.89382          2.767984   
2021-10-01                    1013514          2.074277          2.781712   
2021-09-01                     856562          2.031582          2.766923   
...                               ...               ...               ...   
2005-05-01                     714859          2.394665          2.601995   
2005-04-01                     608493           2.24809          2.675847   
2005-03-01                     545853          2.273948           2.60359   
2005-02-01                     485152          2.267152          2.630021   
2005-01-01                     401350          2.034586           2.64488   

            total_viajeros  total_pernoctaciones  media_total_pernoc  
periodo                                                               
2022-01-01          301474                745224            2.471935  
2021-12-01          394432                957028            2.426345  
2021-11-01          435425               1099696            2.525569  
2021-10-01          494753               1284008            2.595251  
2021-09-01          411625               1063891            2.584612  
...                    ...                   ...                 ...  
2005-05-01          393233                998622            2.539517  
2005-04-01          345617                874251            2.529537  
2005-03-01          308129                769780            2.498239  
2005-02-01          275595                691753            2.510035  
2005-01-01          237561                575948            2.424422  

[205 rows x 9 columns]